processing/src/main/java/org/apache/druid/segment/join/MapJoinableFactory.java

```diff
@@ -19,49 +19,64 @@
 
 package org.apache.druid.segment.join;
 
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.SetMultimap;
 import com.google.inject.Inject;
+import org.apache.druid.java.util.common.ISE;
 import org.apache.druid.query.DataSource;
 
-import java.util.IdentityHashMap;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
 
 /**
  * A {@link JoinableFactory} that delegates to the appropriate factory based on the type of the datasource.
  *
- * Datasources can register a factory via a DruidBinder
+ * Datasources can register a factory via a DruidBinder. Any number of factories can be bound to a datasource, the
+ * 'first' that matches will be returned to the caller, or none if no matches.
  */
 public class MapJoinableFactory implements JoinableFactory
 {
-  private final Map<Class<? extends DataSource>, JoinableFactory> joinableFactories;
+  private final SetMultimap<Class<? extends DataSource>, JoinableFactory> joinableFactories;
 
   @Inject
-  public MapJoinableFactory(Map<Class<? extends DataSource>, JoinableFactory> joinableFactories)
+  public MapJoinableFactory(
+      Set<JoinableFactory> factories,
+      Map<Class<? extends JoinableFactory>, Class<? extends DataSource>> factoryToDataSource
+  )
   {
-    // Accesses to IdentityHashMap should be faster than to HashMap or ImmutableMap.
-    // Class doesn't override Object.equals().
-    this.joinableFactories = new IdentityHashMap<>(joinableFactories);
+    this.joinableFactories = HashMultimap.create();
+    factories.forEach(joinableFactory -> {
+      joinableFactories.put(factoryToDataSource.get(joinableFactory.getClass()), joinableFactory);
+    });
   }
 
   @Override
   public boolean isDirectlyJoinable(DataSource dataSource)
   {
-    JoinableFactory factory = joinableFactories.get(dataSource.getClass());
-    if (factory == null) {
-      return false;
-    } else {
-      return factory.isDirectlyJoinable(dataSource);
+    Set<JoinableFactory> factories = joinableFactories.get(dataSource.getClass());
+    for (JoinableFactory factory : factories) {
+      if (factory.isDirectlyJoinable(dataSource)) {
+        return true;
+      }
     }
+    return false;
   }
 
   @Override
   public Optional<Joinable> build(DataSource dataSource, JoinConditionAnalysis condition)
   {
-    JoinableFactory factory = joinableFactories.get(dataSource.getClass());
-    if (factory == null) {
-      return Optional.empty();
-    } else {
-      return factory.build(dataSource, condition);
+    Set<JoinableFactory> factories = joinableFactories.get(dataSource.getClass());
+    Optional<Joinable> maybeJoinable = Optional.empty();
+    for (JoinableFactory factory : factories) {
+      Optional<Joinable> candidate = factory.build(dataSource, condition);
+      if (candidate.isPresent()) {
+        if (maybeJoinable.isPresent()) {
+          throw new ISE("Multiple joinable factories are valid for table[%s]", dataSource);
```

Comment:
Can you explain why this is an error condition? From the comment at https://github.com/apache/druid/pull/10224/files#diff-51b09b44a8330d19728379390eef60a0R35 it sounds like multiple valid should be okay

Reply:
That comment isn't super clear, will add some more details. Multiple factories can be bound to the same class of datasource (because factory is an extension point), but not to the same exact datasource (because only 1 factory can create the joinable to use in the query). Having a mix of joinables in the same datasource is an invalid state, so this explodes to prevent undefined behavior instead of just using whichever is first in the set.